proxy/round_tripper/proxy_round_tripper.go

```diff
@@ -233,13 +236,39 @@ func (rt *roundTripper) backendRoundTrip(
 
 	rt.combinedReporter.CaptureRoutingRequest(endpoint)
 	tr := GetRoundTripper(endpoint, rt.roundTripperFactory)
-	res, err := tr.RoundTrip(request)
+	res, err := rt.timedRoundTrip(tr, request)
 
 	// decrement connection stats
 	iter.PostRequest(endpoint)
 	return res, err
 }
 
+func (rt *roundTripper) timedRoundTrip(tr http.RoundTripper, request *http.Request) (*http.Response, error) {
+
+	if rt.endpointTimeout <= 0 {
+		return tr.RoundTrip(request)
+	}
+	reqCtx, cancel := context.WithTimeout(request.Context(), rt.endpointTimeout)
+	request = request.WithContext(reqCtx)
+	errored := make(chan struct{}, 1)
```

Comment:
I don't believe this `errored` channel is going to be necessary.

Reply:
We need the errored channel to be able to stop the go routine in the situation that we have canceled the context. In that situation we will be waiting on the channel from `ctx.Done()` indefinitely meaning we would start leaking goroutines.